src/lib/storage/table.cpp

```diff
@@ -122,8 +132,29 @@ std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const {
 }
 
 void Table::compress_chunk(const ChunkID chunk_id) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  auto compressed_chunk = std::make_shared<Chunk>();
+  auto compressed_segments = std::vector<std::shared_ptr<AbstractSegment>>(column_count());
+  auto threads = std::vector<std::thread>(column_count());
+
+  const auto compression_functor = [&](ColumnID column_id) {
+    resolve_data_type(_column_types[column_id], [&](const auto data_type_t) {
+      using ColumnDataType = typename decltype(data_type_t)::type;
+
+      auto value_segment = get_chunk(chunk_id)->get_segment(column_id);
+      auto compressed_segment = std::make_shared<DictionarySegment<ColumnDataType>>(value_segment);
+      compressed_segments[column_id] = compressed_segment;
+    });
+  };
+
+  for (auto col_id = ColumnID{0}; col_id < column_count(); ++col_id) {
+    threads[col_id] = std::thread(compression_functor, col_id);
+  }
+
+  for (auto thread_index = uint32_t{0}; thread_index < threads.size(); ++thread_index) {
+    threads[thread_index].join();
+    compressed_chunk->add_segment(compressed_segments[thread_index]);
+  }
+  _chunks[chunk_id] = compressed_chunk;
```

Comment:
```suggestion
   std::atomic_store(&_chunks.at(chunk_id), compressed_chunk);
```
We used `atomic_store` in our solution. Considering concurrent accesses updating the `shared_ptr` like this could result in unwanted behavior.

Reply:
At this point, there is no more concurrent access anymore, since all this threads are already joined. Do you refer to multithreading the whole database?

Reply:
Yes, I was referring to the whole database and not the multithreading of the threads. So for instance multiple clients accessing the  `_chunks[chunk_id]` while you update it in line 157.